### PR TITLE
[FLINK-16692] flink joblistener can register from config

### PR DESCRIPTION
**Pull request by liuyongvs**
## What is the purpose of the change
- user can implement  JobListener and register by conf. it will be convinient for users to port with other system ,such as atlas

## Verifying this change
- i have unit test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes**)
  - The serializers: no )
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by kl0u** [comment]
Hi @liuyongvs . Thanks for the work, I had some comments on the JIRA and I will have a better look at the PR now. In the meantime, could you make your changes as a single commit? Although not forced yet, we do not allow merge commits in the repo.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 1]
```diff
@@ -237,6 +237,13 @@ public ExecutionConfig getConfig() {
 		return cacheFile;
 	}
 
+	/**
+	 * Gets the config JobListeners.
+	 */
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: Annotate as `@PublicEvolving`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 2]
```diff
@@ -761,6 +768,20 @@ public void configure(ReadableConfig configuration, ClassLoader classLoader) {
 			});
 		config.configure(configuration, classLoader);
 		checkpointCfg.configure(configuration);
+		configuration.getOptional(ExecutionOptions.JOB_LISTENERS)
+			.ifPresent(l -> {
+				for (String listener : l) {
+					try {
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: You can use  the method `InstantiationUtil#instantiate(final String className, final Class<T> targetType, final ClassLoader classLoader)`.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironmentComplexConfigurationTest.java [hunk 3]
```diff
@@ -97,4 +105,47 @@ public void testNotOverridingCachedFilesFromConfiguration() {
...
 	}
+
+	@Test
+	public void testLoadingListenersFromConfiguration() {
+		StreamExecutionEnvironment envFromConfiguration = StreamExecutionEnvironment.getExecutionEnvironment();
+		List<Class> listenersClass = Arrays.asList(BasicJobSubmittedCounter.class, BasicJobExecutedCounter.class);
+
+		Configuration configuration = new Configuration();
+		configuration.setString("execution.job-listeners", listenersClass.stream().map(l -> l.getName()).collect(Collectors.joining(";")));
+
```
Comment (kl0u) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironmentComplexConfigurationTest.java]: You can use the `ConfigUtils#encodeCollectionToConfig()`.

File: flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java [hunk 4]
```diff
@@ -55,4 +56,11 @@
 						"throughput")
 				)
 				.build());
+
+	public static final ConfigOption<List<String>> JOB_LISTENERS =
```
Comment (kl0u) [inline flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java]: I would suggest to move it to the `DeploymentOptions`.

**Comment by liuyongvs** [comment]
Thanks you so much. Becase my personal PC doen't run fast. so i just not run on my PC and commits directly. i will commit with one commits next time.
> Hi @liuyongvs . Thanks for the work, I had some comments on the JIRA and I will have a better look at the PR now.
> 
> In the meantime, could you make your changes as a single commit and rename your branch? Although not forced yet, we do not allow merge commits in the repo.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 1 above)
Comment (liuyongvs) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: thanks, but ExecutionEnvironment is for Dataset api not DatatStream Api.

Comment (kl0u) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: Yes, `ExecutionEnvironment` is for the `DataSet` API and there are also `JobListeners` there. So why not adding it?

File: flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java (hunk 4 above)
Comment (liuyongvs) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java]: i have moved the argument from ExecutionOptions to   DeploymentOptions.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 2 above)
Comment (liuyongvs) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: I have used this tool

**Comment by liuyongvs** [comment]
Hi @kl0u , it is ok now

**Comment by kl0u** [comment]
Hi @liuyongvs there is no need to run on your PC in order to create a new branch. You can just go to a clean master branch on your laptop, create a new branch, cherry pick all the commits from this PR, squash them and push to github :)

**Comment by kl0u** [comment]
@liuyongvs I am reviewing currently your PR. This is the branch https://github.com/kl0u/flink/tree/listeners that I created based on your commits. Please check if all your changes are included, and if I have any comments, please integrate them on that branch.

**Comment by kl0u** [comment]
Changes look good, I will merge as soon as my travis gives green. Thanks for the work @liuyongvs !

**Comment by liuyongvs** [comment]
Thanks @kl0u , I read the commits , all my changes are included
> @liuyongvs I am reviewing currently your PR. This is the branch https://github.com/kl0u/flink/tree/listeners that I created based on your commits. Please check if all your changes are included, and if I have any comments, please integrate them on that branch.

**Pull request closed without merging**